# Pandas Dataframe

<img src="https://frenzy86.s3.eu-west-2.amazonaws.com/IFAO/dataframe.png" width="1200">

In [1]:
import numpy as np
import pandas as pd

In [2]:
np.random.seed(667)
mydata = np.random.randint(0,101,(4,3))

In [3]:
mydata

array([[ 85,  15,  89],
       [100,   3,  90],
       [  2,  64,   0],
       [ 36,  99,  47]])

In [4]:
myindex = ['CA','NY','AZ','TX']

In [5]:
mycolumns = ['Jan','Feb','Mar']

In [6]:
pd.DataFrame()

Empty DataFrame
Columns: []
Index: []

In [7]:
df = pd.DataFrame(data=mydata)
df

0   1   2
0   85  15  89
1  100   3  90
2    2  64   0
3   36  99  47

In [8]:
df = pd.DataFrame(index=myindex,data=mydata)
df

0   1   2
CA   85  15  89
NY  100   3  90
AZ    2  64   0
TX   36  99  47

In [9]:
# df = pd.DataFrame(myindex,mycolumns,mydata)


In [10]:
df = pd.DataFrame(data=mydata,index=myindex,columns=mycolumns)
df

Jan  Feb  Mar
CA   85   15   89
NY  100    3   90
AZ    2   64    0
TX   36   99   47

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4 entries, CA to TX
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   Jan     4 non-null      int64
 1   Feb     4 non-null      int64
 2   Mar     4 non-null      int64
dtypes: int64(3)
memory usage: 128.0+ bytes


In [12]:
df['Jan']

CA     85
NY    100
AZ      2
TX     36
Name: Jan, dtype: int64

In [13]:
df['Jan'].tolist()

[85, 100, 2, 36]

In [14]:
df['Jan'].values

array([ 85, 100,   2,  36])

In [15]:
data = {'Nome':['Gianni','Enrico','Sergio','Andrea','Marco'],
        'Eta': [23,56,32,None,35],
        'Professione':['cuoco','postino','manager','sbirro','inoccupato']
        }

In [16]:
df1 = pd.DataFrame(data)
df1

Nome   Eta Professione
0  Gianni  23.0       cuoco
1  Enrico  56.0     postino
2  Sergio  32.0     manager
3  Andrea   NaN      sbirro
4   Marco  35.0  inoccupato

In [17]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Nome         5 non-null      object 
 1   Eta          4 non-null      float64
 2   Professione  5 non-null      object 
dtypes: float64(1), object(2)
memory usage: 248.0+ bytes


In [18]:
df1.describe().T

count  mean       std   min    25%   50%    75%   max
Eta    4.0  36.5  13.96424  23.0  29.75  33.5  40.25  56.0

## Read external data tables

In [19]:
!wget https://frenzy86.s3.eu-west-2.amazonaws.com/IFAO/tips_complete.csv

--2024-04-15 10:11:54--  https://frenzy86.s3.eu-west-2.amazonaws.com/IFAO/tips_complete.csv
Resolving frenzy86.s3.eu-west-2.amazonaws.com (frenzy86.s3.eu-west-2.amazonaws.com)... 52.95.150.146, 52.95.149.50, 52.95.191.6, ...
Connecting to frenzy86.s3.eu-west-2.amazonaws.com (frenzy86.s3.eu-west-2.amazonaws.com)|52.95.150.146|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18752 (18K) [text/csv]
Saving to: ‘tips_complete.csv.3’

tips_complete.csv.3 100%[===================>]  18.31K  --.-KB/s    in 0.001s  

2024-04-15 10:11:54 (21.8 MB/s) - ‘tips_complete.csv.3’ saved [18752/18752]



In [74]:
ls

sample_data/  tips_complete.csv  tips_complete.csv.1  tips_complete.csv.2  tips_complete.csv.3


In [75]:
df = pd.read_csv('tips_complete.csv')
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  
0    Christy Cunningham  3560325168603410    Sun2959  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
4          Tonya Carter  4832732618637221    Sun2251  
..                  ...               ...        ...  
239       Michael Avila  5296068606052842    Sat2657  
240      Monica Sanders  3506806155565404    Sat1766  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  
243     Michelle Hardin  3511451626698139    Thur672  

[244 rows x 11 columns]

----
About this DataSet (in case you are interested)

* Description
    * One waiter recorded information about each tip he received over a period of a few months working in one restaurant. He collected several variables:

* Format
    * A data frame with 244 rows and 7 variables

* Details
    * tip in dollars,
    * bill in dollars,
    * sex of the bill payer,
    * whether there were smokers in the party,
    * day of the week,
    * time of day,
    * size of the party.

In all he recorded 244 tips.
    
* Note: I created some additional columns with Fake data, including Name, CC Number, and Payment ID.

----

# DataFrames

In [22]:
df.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size',
       'price_per_person', 'Payer Name', 'CC Number', 'Payment ID'],
      dtype='object')

In [23]:
df.index

RangeIndex(start=0, stop=244, step=1)

In [76]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

In [77]:
df.tail(3)

total_bill   tip     sex smoker   day    time  size  price_per_person  \
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

          Payer Name         CC Number Payment ID  
241       Keith Wong  6011891618747196    Sat3880  
242     Dennis Dixon     4375220550950      Sat17  
243  Michelle Hardin  3511451626698139    Thur672

In [78]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   total_bill        244 non-null    float64
 1   tip               244 non-null    float64
 2   sex               244 non-null    object 
 3   smoker            244 non-null    object 
 4   day               244 non-null    object 
 5   time              244 non-null    object 
 6   size              244 non-null    int64  
 7   price_per_person  244 non-null    float64
 8   Payer Name        244 non-null    object 
 9   CC Number         244 non-null    int64  
 10  Payment ID        244 non-null    object 
dtypes: float64(3), int64(2), object(6)
memory usage: 21.1+ KB


In [79]:
len(df)#solo numero di righe

244

In [80]:
df.shape

(244, 11)

In [82]:
df.describe()

total_bill         tip        size  price_per_person     CC Number
count  244.000000  244.000000  244.000000        244.000000  2.440000e+02
mean    19.785943    2.998279    2.569672          7.888197  2.563496e+15
std      8.902412    1.383638    0.951100          2.914234  2.369340e+15
min      3.070000    1.000000    1.000000          2.880000  6.040679e+10
25%     13.347500    2.000000    2.000000          5.800000  3.040731e+13
50%     17.795000    2.900000    2.000000          7.255000  3.525318e+15
75%     24.127500    3.562500    3.000000          9.390000  4.553675e+15
max     50.810000   10.000000    6.000000         20.270000  6.596454e+15

In [30]:
#df.describe().transpose()
df.describe().T

count          mean           std           min  \
total_bill        244.0  1.978594e+01  8.902412e+00  3.070000e+00   
tip               244.0  2.998279e+00  1.383638e+00  1.000000e+00   
size              244.0  2.569672e+00  9.510998e-01  1.000000e+00   
price_per_person  244.0  7.888197e+00  2.914234e+00  2.880000e+00   
CC Number         244.0  2.563496e+15  2.369340e+15  6.040679e+10   

                           25%           50%           75%           max  
total_bill        1.334750e+01  1.779500e+01  2.412750e+01  5.081000e+01  
tip               2.000000e+00  2.900000e+00  3.562500e+00  1.000000e+01  
size              2.000000e+00  2.000000e+00  3.000000e+00  6.000000e+00  
price_per_person  5.800000e+00  7.255000e+00  9.390000e+00  2.027000e+01  
CC Number         3.040731e+13  3.525318e+15  4.553675e+15  6.596454e+15

In [31]:
df.sum()

total_bill                                                    4827.77
tip                                                            731.58
sex                 FemaleMaleMaleMaleFemaleMaleMaleMaleMaleMaleMa...
smoker              NoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNo...
day                 SunSunSunSunSunSunSunSunSunSunSunSunSunSunSunS...
time                DinnerDinnerDinnerDinnerDinnerDinnerDinnerDinn...
size                                                              627
price_per_person                                              1924.72
Payer Name          Christy CunninghamDouglas TuckerTravis Walters...
CC Number                                          625492917132785482
Payment ID          Sun2959Sun4608Sun4458Sun5260Sun2251Sun9679Sun5...
dtype: object

## Selection and Indexing

Let's learn how to retrieve information from a DataFrame.

### COLUMNS

We will begin be learning how to extract information based on the columns

In [32]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

#### Grab a Single Column

In [33]:
df['total_bill']

0      16.99
1      10.34
2      21.01
3      23.68
4      24.59
       ...  
239    29.03
240    27.18
241    22.67
242    17.82
243    18.78
Name: total_bill, Length: 244, dtype: float64

In [34]:
type(df['total_bill'])

pandas.core.series.Series

In [35]:
df[['tip','total_bill']].mean()

tip            2.998279
total_bill    19.785943
dtype: float64

In [36]:
df.total_bill

0      16.99
1      10.34
2      21.01
3      23.68
4      24.59
       ...  
239    29.03
240    27.18
241    22.67
242    17.82
243    18.78
Name: total_bill, Length: 244, dtype: float64

#### Grab Multiple Columns

In [83]:
# Note how its a python list of column names! Thus the double brackets.
df[['total_bill','tip']] #è una lista di selezione!!!

total_bill   tip
0         16.99  1.01
1         10.34  1.66
2         21.01  3.50
3         23.68  3.31
4         24.59  3.61
..          ...   ...
239       29.03  5.92
240       27.18  2.00
241       22.67  2.00
242       17.82  1.75
243       18.78  3.00

[244 rows x 2 columns]

#### Create New Columns

In [84]:
df['new'] = 2
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  new  
0    Christy Cunningham  3560325168603410    Sun2959    2  
1        Douglas Tucker  4478071379779230    Sun4608    2  
2        Travis Walters  6011812112971322    Sun4458    2  
3      Nathaniel Harris  4676137647685994    Sun5260    2  
4          Tonya Carter  4832732618637221    Sun2251    2  
..                  ...               ...        ...  ...  
239       Michael Avila  5296068606052842    Sat2657    2  
240      Monica Sanders  3506806155565404    Sat1766    2  
241          Keith Wong  6011891618747196    Sat3880    2  
242        Dennis Dixon     4375220550950      Sat17    2  
243     Michelle Hardin  3511451626698139    Thur672    2  

[244 rows x 12 columns]

In [86]:
100* df['tip'] / df['total_bill']

0       5.944673
1      16.054159
2      16.658734
3      13.978041
4      14.680765
         ...    
239    20.392697
240     7.358352
241     8.822232
242     9.820426
243    15.974441
Length: 244, dtype: float64

In [87]:
df['tip_percentage'] = 100* df['tip'] / df['total_bill']

In [88]:
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  new  tip_percentage  
0    Christy Cunningham  3560325168603410    Sun2959    2        5.944673  
1        Douglas Tucker  4478071379779230    Sun4608    2       16.054159  
2        Travis Walters  6011812112971322    Sun4458    2       16.658734  
3      Nathaniel Harris  4676137647685994    Sun5260    2       13.978041  
4          Tonya Carter  4832732618637221    Sun2251    2       14.680765  
..                  ...               ...        ...  ...             ...  
239       Michael Avila  5296068606052842    Sat2657    2       20.392697  
240      Monica Sanders  3506806155565404    Sat1766    2        7.358352  
241          Keith Wong  6011891618747196    Sat3880    2        8.822232  
242        Dennis Dixon     4375220550950      Sat17    2        9.820426  
243     Michelle Hardin  3511451626698139    Thur672    2       15.974441  

[244 rows x 13 columns]

In [91]:
df['price_per_person'] = df['total_bill'] / df['size']

In [92]:
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2          8.495000   
1         10.34  1.66    Male     No   Sun  Dinner     3          3.446667   
2         21.01  3.50    Male     No   Sun  Dinner     3          7.003333   
3         23.68  3.31    Male     No   Sun  Dinner     2         11.840000   
4         24.59  3.61  Female     No   Sun  Dinner     4          6.147500   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3          9.676667   
240       27.18  2.00  Female    Yes   Sat  Dinner     2         13.590000   
241       22.67  2.00    Male    Yes   Sat  Dinner     2         11.335000   
242       17.82  1.75    Male     No   Sat  Dinner     2          8.910000   
243       18.78  3.00  Female     No  Thur  Dinner     2          9.390000   

             Payer Name         CC Number Payment ID  new  tip_percentage  
0    Christy Cunningham  3560325168603410    Sun2959    2        5.944673  
1        Douglas Tucker  4478071379779230    Sun4608    2       16.054159  
2        Travis Walters  6011812112971322    Sun4458    2       16.658734  
3      Nathaniel Harris  4676137647685994    Sun5260    2       13.978041  
4          Tonya Carter  4832732618637221    Sun2251    2       14.680765  
..                  ...               ...        ...  ...             ...  
239       Michael Avila  5296068606052842    Sat2657    2       20.392697  
240      Monica Sanders  3506806155565404    Sat1766    2        7.358352  
241          Keith Wong  6011891618747196    Sat3880    2        8.822232  
242        Dennis Dixon     4375220550950      Sat17    2        9.820426  
243     Michelle Hardin  3511451626698139    Thur672    2       15.974441  

[244 rows x 13 columns]

In [93]:
help(np.round)

Help on _ArrayFunctionDispatcher in module numpy:

round(a, decimals=0, out=None)
    Evenly round to the given number of decimals.
    
    Parameters
    ----------
    a : array_like
        Input data.
    decimals : int, optional
        Number of decimal places to round to (default: 0).  If
        decimals is negative, it specifies the number of positions to
        the left of the decimal point.
    out : ndarray, optional
        Alternative output array in which to place the result. It must have
        the same shape as the expected output, but the type of the output
        values will be cast if necessary. See :ref:`ufuncs-output-type` for more
        details.
    
    Returns
    -------
    rounded_array : ndarray
        An array of the same type as `a`, containing the rounded values.
        Unless `out` was specified, a new array is created.  A reference to
        the result is returned.
    
        The real and imaginary parts of complex numbers are rounded
      

#### Adjust Existing Columns

In [94]:
# Because pandas is based on numpy, we get awesome capabilities with numpy's universal functions!
df['price_per_person'] = np.round(df['price_per_person'],2)

In [95]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  new  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959    2        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608    2       16.054159  
2      Travis Walters  6011812112971322    Sun4458    2       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260    2       13.978041  
4        Tonya Carter  4832732618637221    Sun2251    2       14.680765

#### Remove Columns

In [96]:
df.drop(['tip_percentage'],axis=1)

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  new  
0    Christy Cunningham  3560325168603410    Sun2959    2  
1        Douglas Tucker  4478071379779230    Sun4608    2  
2        Travis Walters  6011812112971322    Sun4458    2  
3      Nathaniel Harris  4676137647685994    Sun5260    2  
4          Tonya Carter  4832732618637221    Sun2251    2  
..                  ...               ...        ...  ...  
239       Michael Avila  5296068606052842    Sat2657    2  
240      Monica Sanders  3506806155565404    Sat1766    2  
241          Keith Wong  6011891618747196    Sat3880    2  
242        Dennis Dixon     4375220550950      Sat17    2  
243     Michelle Hardin  3511451626698139    Thur672    2  

[244 rows x 12 columns]

In [117]:
df

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sun2959          16.99  1.01  Female     No   Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No   Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No   Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No   Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No   Sun  Dinner     4   
...                ...   ...     ...    ...   ...     ...   ...   
Sat2657          29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person          Payer Name         CC Number  new  
Payment ID                                                               
Sun2959                 8.49  Christy Cunningham  3560325168603410    2  
Sun4608                 3.45      Douglas Tucker  4478071379779230    2  
Sun4458                 7.00      Travis Walters  6011812112971322    2  
Sun5260                11.84    Nathaniel Harris  4676137647685994    2  
Sun2251                 6.15        Tonya Carter  4832732618637221    2  
...                      ...                 ...               ...  ...  
Sat2657                 9.68       Michael Avila  5296068606052842    2  
Sat1766                13.59      Monica Sanders  3506806155565404    2  
Sat3880                11.34          Keith Wong  6011891618747196    2  
Sat17                   8.91        Dennis Dixon     4375220550950    2  
Thur672                 9.39     Michelle Hardin  3511451626698139    2  

[244 rows x 11 columns]

In [98]:
df = df.drop("tip_percentage",axis=1) #sovrascrivere!!!!!!!

In [99]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  new  
0  Christy Cunningham  3560325168603410    Sun2959    2  
1      Douglas Tucker  4478071379779230    Sun4608    2  
2      Travis Walters  6011812112971322    Sun4458    2  
3    Nathaniel Harris  4676137647685994    Sun5260    2  
4        Tonya Carter  4832732618637221    Sun2251    2

# Index Basics

Before going over the same retrieval tasks for rows, let's build some basic understanding of the pandas DataFrame Index.

In [100]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  new  
0  Christy Cunningham  3560325168603410    Sun2959    2  
1      Douglas Tucker  4478071379779230    Sun4608    2  
2      Travis Walters  6011812112971322    Sun4458    2  
3    Nathaniel Harris  4676137647685994    Sun5260    2  
4        Tonya Carter  4832732618637221    Sun2251    2

In [101]:
df.index

RangeIndex(start=0, stop=244, step=1)

In [102]:
df.set_index('Payment ID')

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sun2959          16.99  1.01  Female     No   Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No   Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No   Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No   Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No   Sun  Dinner     4   
...                ...   ...     ...    ...   ...     ...   ...   
Sat2657          29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person          Payer Name         CC Number  new  
Payment ID                                                               
Sun2959                 8.49  Christy Cunningham  3560325168603410    2  
Sun4608                 3.45      Douglas Tucker  4478071379779230    2  
Sun4458                 7.00      Travis Walters  6011812112971322    2  
Sun5260                11.84    Nathaniel Harris  4676137647685994    2  
Sun2251                 6.15        Tonya Carter  4832732618637221    2  
...                      ...                 ...               ...  ...  
Sat2657                 9.68       Michael Avila  5296068606052842    2  
Sat1766                13.59      Monica Sanders  3506806155565404    2  
Sat3880                11.34          Keith Wong  6011891618747196    2  
Sat17                   8.91        Dennis Dixon     4375220550950    2  
Thur672                 9.39     Michelle Hardin  3511451626698139    2  

[244 rows x 11 columns]

In [103]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  new  
0  Christy Cunningham  3560325168603410    Sun2959    2  
1      Douglas Tucker  4478071379779230    Sun4608    2  
2      Travis Walters  6011812112971322    Sun4458    2  
3    Nathaniel Harris  4676137647685994    Sun5260    2  
4        Tonya Carter  4832732618637221    Sun2251    2

In [104]:
df = df.set_index('Payment ID')#sovrascrivere!!!!!!!
df

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sun2959          16.99  1.01  Female     No   Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No   Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No   Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No   Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No   Sun  Dinner     4   
...                ...   ...     ...    ...   ...     ...   ...   
Sat2657          29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person          Payer Name         CC Number  new  
Payment ID                                                               
Sun2959                 8.49  Christy Cunningham  3560325168603410    2  
Sun4608                 3.45      Douglas Tucker  4478071379779230    2  
Sun4458                 7.00      Travis Walters  6011812112971322    2  
Sun5260                11.84    Nathaniel Harris  4676137647685994    2  
Sun2251                 6.15        Tonya Carter  4832732618637221    2  
...                      ...                 ...               ...  ...  
Sat2657                 9.68       Michael Avila  5296068606052842    2  
Sat1766                13.59      Monica Sanders  3506806155565404    2  
Sat3880                11.34          Keith Wong  6011891618747196    2  
Sat17                   8.91        Dennis Dixon     4375220550950    2  
Thur672                 9.39     Michelle Hardin  3511451626698139    2  

[244 rows x 11 columns]

In [56]:
# df = df.set_index('tip')#sovrascrivere!!!!!!!
# df

In [105]:
df.index

Index(['Sun2959', 'Sun4608', 'Sun4458', 'Sun5260', 'Sun2251', 'Sun9679',
       'Sun5985', 'Sun8157', 'Sun6820', 'Sun3775',
       ...
       'Sat7220', 'Sat4615', 'Sat5032', 'Sat2929', 'Sat9777', 'Sat2657',
       'Sat1766', 'Sat3880', 'Sat17', 'Thur672'],
      dtype='object', name='Payment ID', length=244)

### ROWS

Let's now explore these same concepts but with Rows.

#### Grab a Single Row

In [107]:
# Integer Based
df.iloc[0] # prende la riga con indice indicato

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
new                                  2
Name: Sun2959, dtype: object

In [110]:
# Name Based
#df = df.set_index('Payment ID')
df.loc['Sun2959'] # pesca dall'index

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
new                                  2
Name: Sun2959, dtype: object

#### Grab Multiple Rows

In [59]:
df.iloc[0:4] #prime 4 righe

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  new  
Payment ID                                                               
Sun2959                 8.49  Christy Cunningham  3560325168603410    2  
Sun4608                 3.45      Douglas Tucker  4478071379779230    2  
Sun4458                 7.00      Travis Walters  6011812112971322    2  
Sun5260                11.84    Nathaniel Harris  4676137647685994    2

In [60]:
df.loc[['Sun2959','Sun5260']]

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  new  
Payment ID                                                               
Sun2959                 8.49  Christy Cunningham  3560325168603410    2  
Sun5260                11.84    Nathaniel Harris  4676137647685994    2

In [61]:
df.iloc[:,-2:]

CC Number  new
Payment ID                       
Sun2959     3560325168603410    2
Sun4608     4478071379779230    2
Sun4458     6011812112971322    2
Sun5260     4676137647685994    2
Sun2251     4832732618637221    2
...                      ...  ...
Sat2657     5296068606052842    2
Sat1766     3506806155565404    2
Sat3880     6011891618747196    2
Sat17          4375220550950    2
Thur672     3511451626698139    2

[244 rows x 2 columns]

#### Remove Row

Typically are datasets will be large enough that we won't remove rows like this since we won't know thier row location for some specific condition, instead, we drop rows based on conditions such as missing data or column values. The next lecture will cover this in a lot more detail.

In [62]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  new  
Payment ID                                                               
Sun2959                 8.49  Christy Cunningham  3560325168603410    2  
Sun4608                 3.45      Douglas Tucker  4478071379779230    2  
Sun4458                 7.00      Travis Walters  6011812112971322    2  
Sun5260                11.84    Nathaniel Harris  4676137647685994    2  
Sun2251                 6.15        Tonya Carter  4832732618637221    2

In [111]:
df.drop('Sun2959',axis=0).head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   
Sun9679          25.29  4.71    Male     No  Sun  Dinner     4   

            price_per_person        Payer Name         CC Number  new  
Payment ID                                                             
Sun4608                 3.45    Douglas Tucker  4478071379779230    2  
Sun4458                 7.00    Travis Walters  6011812112971322    2  
Sun5260                11.84  Nathaniel Harris  4676137647685994    2  
Sun2251                 6.15      Tonya Carter  4832732618637221    2  
Sun9679                 6.32        Erik Smith   213140353657882    2

In [64]:
# Error if you have a named index!
# df.drop(0,axis=0).head()

#### Insert a New Row

Pretty rare to add a single row like this. Usually you use pd.concat() to add many rows at once. You could use the .append() method with a list of pd.Series() objects, but you won't see us do this with realistic real-world data.

In [112]:
one_row = df.iloc[0]

In [113]:
one_row

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
new                                  2
Name: Sun2959, dtype: object

In [114]:
type(one_row)

pandas.core.series.Series

In [115]:
df.tail()

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sat2657          29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person       Payer Name         CC Number  new  
Payment ID                                                            
Sat2657                 9.68    Michael Avila  5296068606052842    2  
Sat1766                13.59   Monica Sanders  3506806155565404    2  
Sat3880                11.34       Keith Wong  6011891618747196    2  
Sat17                   8.91     Dennis Dixon     4375220550950    2  
Thur672                 9.39  Michelle Hardin  3511451626698139    2

In [116]:
df.append(one_row).tail() ## deprecato

AttributeError: 'DataFrame' object has no attribute 'append'

--------

#### Extract an array

In [70]:
array_row = df.iloc[0].values
array_row

array([16.99, 1.01, 'Female', 'No', 'Sun', 'Dinner', 2, 8.49,
       'Christy Cunningham', 3560325168603410, 2], dtype=object)

In [71]:
df['total_bill'].tolist()

[16.99,
 10.34,
 21.01,
 23.68,
 24.59,
 25.29,
 8.77,
 26.88,
 15.04,
 14.78,
 10.27,
 35.26,
 15.42,
 18.43,
 14.83,
 21.58,
 10.33,
 16.29,
 16.97,
 20.65,
 17.92,
 20.29,
 15.77,
 39.42,
 19.82,
 17.81,
 13.37,
 12.69,
 21.7,
 19.65,
 9.55,
 18.35,
 15.06,
 20.69,
 17.78,
 24.06,
 16.31,
 16.93,
 18.69,
 31.27,
 16.04,
 17.46,
 13.94,
 9.68,
 30.4,
 18.29,
 22.23,
 32.4,
 28.55,
 18.04,
 12.54,
 10.29,
 34.81,
 9.94,
 25.56,
 19.49,
 38.01,
 26.41,
 11.24,
 48.27,
 20.29,
 13.81,
 11.02,
 18.29,
 17.59,
 20.08,
 16.45,
 3.07,
 20.23,
 15.01,
 12.02,
 17.07,
 26.86,
 25.28,
 14.73,
 10.51,
 17.92,
 27.2,
 22.76,
 17.29,
 19.44,
 16.66,
 10.07,
 32.68,
 15.98,
 34.83,
 13.03,
 18.28,
 24.71,
 21.16,
 28.97,
 22.49,
 5.75,
 16.32,
 22.75,
 40.17,
 27.28,
 12.03,
 21.01,
 12.46,
 11.35,
 15.38,
 44.3,
 22.42,
 20.92,
 15.36,
 20.49,
 25.21,
 18.24,
 14.31,
 14.0,
 7.25,
 38.07,
 23.95,
 25.71,
 17.31,
 29.93,
 10.65,
 12.43,
 24.08,
 11.69,
 13.42,
 14.26,
 15.95,
 12.48,
 29.8,
 8.52,